src/rewrites.rs

```diff
@@ -588,3 +712,274 @@ fn check_pat(
+                        }
+                    }
+                }
+            }
+        }
+
+        runner.egraph.rebuild();
+        Ok(())
+    }
+}
+
+/// Canonicalize a pattern
+///
+/// This function constructs a canonicalized pattern for a given pattern. It sequentially
+/// replaces each unique variable with ?i_0, ?i_1 ...
+///
+/// # Parameters
+///
+/// - `pat`: the pattern to canonicalize. See egg::Pattern for more info
+///
+/// # Returns
+///
+/// - Pattern<Mdl>: the canonicalized pattern
+/// - HashMap<egg::Var, egg::Var>: Mapping from variable in the original pattern to
+///     variable in the canonical pattern.
+fn canonicalize(pat: &Pattern<Mdl>) -> (Pattern<Mdl>, HashMap<egg::Var, egg::Var>) {
+    let mut var_map = HashMap::<egg::Var, egg::Var>::new();
+    let mut count = 0;
+    let substituted: Vec<_> = pat
+        .ast
+        .as_ref()
+        .iter()
+        .cloned()
+        .map(|x| match x {
+            ENodeOrVar::ENode(_) => x,
+            ENodeOrVar::Var(v) => match var_map.get(&v) {
+                Some(var) => ENodeOrVar::Var(*var),
+                None => {
+                    let name = format!("?i_{}", count);
+                    count += 1;
+                    let new_var: egg::Var = name.parse().unwrap();
+                    let new_node = ENodeOrVar::Var(new_var);
+                    var_map.insert(v, new_var);
+                    new_node
+                }
+            },
+        })
+        .collect();
+
+    let ast = RecExpr::from(substituted);
+    let canonical_pat = Pattern::<Mdl>::from(ast);
+    (canonical_pat, var_map)
+}
+
+/// Merge two substitutions
+///
+/// This function merges two substitutions. The merged one contains substitutions (egg::Var -> Id)
+/// from both input substitutions.
+///
+/// # Parameters
+///
+/// - `subst_1`: substitution to be merged
+/// - `subst_2`: substitution to be merged
+/// - `var_map_1`: the keys of this map should be all Var in subst_1. It is just for providing the Vars
+fn merge_subst(
+    subst_1: Subst,
+    mut subst_2: Subst,
+    var_map_1: &HashMap<egg::Var, egg::Var>,
+) -> Subst {
+    for (var, _) in var_map_1.iter() {
+        let id_1 = subst_1.get(*var).unwrap();
+        subst_2.insert(*var, *id_1);
+    }
+    subst_2
+}
+
+/// Decanonicalize a substitution
+///
+/// Create a decanonicalized substitution by replacing the variables in the canonical substitution
+/// with the original variables
+///
+/// # Parameters
+///
+/// - `subst`: The substitution using the canonicalized variables
+/// - `var_map`: Mapping from variable in the original pattern to variable in the canonical pattern.
+fn decanonicalize(subst: &Subst, var_map: &HashMap<egg::Var, egg::Var>) -> Subst {
+    var_map.iter().fold(
+        Default::default(),
+        |mut new_subst, (orig_var, canonical_var)| {
+            new_subst.insert(*orig_var, *subst.get(*canonical_var).unwrap());
+            new_subst
+        },
+    )
```

Comment:
A for-loop might be easier to read here.

Reply:
I remember you recommended using fold before. When is it better to use fold and when it is better to use loops?

Reply:
That's a good question. I do not have a definition answer.

As a rule of thumb, I would:
- use `for` when your action has side effects (e.g. here you need `mut`, so better use a `for`).
- use `fold` when it is purely functional

Reply:
Actually, you can use `collect()` here.
```
var_map.iter().map(|origin, canonical| (*origin, *subst.get(*canonical).unwrap())).collect()
```

Reply:
The returned type is egg::Subst, so I can't really use collect() to construct that. I've changed into using for loop.